### PR TITLE
Improve BlobStore downloads

### PR DESCRIPTION
**Pull request by ddl-gabrielhaim**
- Add new download method for multithreaded downloads
- Stop using hardcoded values
- Release connection at the end of download

## Description

Add new `download` API that utilises multithreaded and range download for optimized speeds

## Related Issue

https://dominodatalab.atlassian.net/browse/DOM-47912

## Type of Change

- [ ] 📚 Examples / docs / tutorials / dependencies update
- [ ] 🔧 Bug fix (non-breaking change which fixes an issue)
- [x] 🥂 Improvement (non-breaking change which improves an existing feature)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🔐 Security fix

## Checklist

- [x] I've read the [`CONTRIBUTING.md`](https://github.com/dominodatalab/domino-data/blob/main/CONTRIBUTING.md) guide.
- [x] I've updated the code style using `make codestyle`.
- [x] I've written tests for all new methods and classes that I created.
- [x] I've written the docstring in Google format for all the methods and classes that I used.

### CODE REVIEW

File: domino_data/data_sources.py [hunk 1]
```diff
@@ -316,6 +341,15 @@ def http(self) -> httpx.Client:
             self._httpx = httpx.Client(verify=context)
         return self._httpx
 
+    def pool_manager(self) -> urllib3.PoolManager:
+        """Urllib3 pool manager for range downloads."""
+        if self.datasource_type == DatasourceDtoDataSourceType.ADLSCONFIG.value:
+            return urllib3.PoolManager(headers=ADLS_HEADERS)
+        elif self.datasource_type == DatasourceDtoDataSourceType.GENERICS3CONFIG.value:
+            return urllib3.PoolManager(assert_hostname=False)
```
Comment (ddl-giuliocapolino) [inline domino_data/data_sources.py:349]: are these things you tested out and figured that they are required given datasource type?

File: domino_data/transfer.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+MB = 2**20
+
+
+def split_range(start: int, end: int, step: int) -> Generator[tuple[int, int], None, None]:
+    """Yield all steps from start to end as tuple of integers.
+
+    All returned block are inclusive to ensure every integer is covered from start to end.
+
+    Example:
+        start: 0, end: 10, step: 3 -> (0, 2), (3, 6), (7, 10)
```
Comment (ddl-giuliocapolino) [inline domino_data/transfer.py:20]: mmh interesting, is it normal that the first tuple in the range has one less element? like 0 -> 2 (0, 1, 2) has 3, and 3 -> 6 (3, 4, 5, 6) has 4

File: domino_data/transfer.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    yield (max_block, end)
+
+
+class BlobTransfer:
+    def __init__(
+        self,
+        url: str,
+        destination: BinaryIO,
+        max_workers: int = MAX_WORKERS,
+        chunk_size: int = 16 * MB,
```
Comment (ddl-giuliocapolino) [inline domino_data/transfer.py:46]: where does the 16 come from?

File: domino_data/transfer.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+import io
+import shutil
+import threading
+from concurrent.futures import ThreadPoolExecutor
+
+import urllib3
+
+MAX_WORKERS = 10
+MB = 2**20
```
Comment (ddl-giuliocapolino) [inline domino_data/transfer.py:11]: mmh maybe a bit more specific of a variable name here?

**Review by ddl-giuliocapolino — APPROVED** [review]
few comments but looks good overall - is there any sort of integration test we can do here or not really from domino-data?

File: domino_data/data_sources.py (hunk 1 above)
Comment (ddl-gabrielhaim) [reply domino_data/data_sources.py:349]: it's the urllib3 equivalent of https://github.com/dominodatalab/domino-data/blob/main/domino_data/data_sources.py#L304

File: domino_data/transfer.py (hunk 4 above)
Comment (ddl-gabrielhaim) [reply domino_data/transfer.py:11]: that's a megabyte 2^20 bytes, I'll make it clearer

File: domino_data/transfer.py (hunk 2 above)
Comment (ddl-gabrielhaim) [reply domino_data/transfer.py:20]: it's actually wrong, I need to change with the values from the actual test. yes the latest block will not always respect the step depending on the modulo of the total size

File: domino_data/transfer.py (hunk 3 above)
Comment (ddl-gabrielhaim) [reply domino_data/transfer.py:46]: https://docs.aws.amazon.com/AmazonS3/latest/userguide/optimizing-performance-guidelines.html#optimizing-performance-guidelines-get-range

**Comment by ronit-basu** [comment]
Will take a look today. Please don't merge before then!

**Comment by ddl-giuliocapolino** [comment]
I think there may be some vulnerability issues also on `requests` and `starlette` that need to be fixed or the build won’t pass

**Pull request merged**